src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/SemanticTokensCache.cs

```diff
@@ -0,0 +1,101 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+#pragma warning disable CS0618
+#nullable enable
+
+using Microsoft.CodeAnalysis.Razor;
+using OmniSharp.Extensions.LanguageServer.Protocol;
+using System.Collections.Generic;
+using System;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models.Proposals;
+using System.Linq;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Semantic
+{
+    /// <summary>
+    /// Cache used by <see cref="DefaultRazorSemanticTokensInfoService"/> to keep track of
+    /// previously computed tokens.
+    /// </summary>
+    internal class SemanticTokensCache<T> where T : SemanticTokens
+    {
+        /// <summary>
+        /// Number of cached token sets we store per document. Must be >= 1.
+        /// </summary>
+        /// <remarks>
+        /// Internal for testing only.
+        /// </remarks>
+        internal const int MaxCachesPerDoc = 5;
+
+        /// <summary>
+        /// Multiple cache requests or updates may be received concurrently. We need this lock to
+        /// ensure that we aren't making concurrent modifications to the _documentToTokenSets
+        /// dictionary.
+        /// </summary>
+        private readonly object _lock = new();
+
+        #region protected by _lock
+        /// <summary>
+        /// Maps a document URI to its n most recently cached token sets. We use an in-memory cache
+        /// to limit memory allocation.
+        /// </summary>
+        private readonly MemoryCache<DocumentUri, List<T>> _documentToTokenSets = new();
```

Comment:
Should we be capping this cache set so we don't capture every document and grow unbounded? Not sure if on the Roslyn side we decided that wasn't necessary